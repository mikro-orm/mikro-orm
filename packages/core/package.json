{
  "name": "@mikro-orm/core",
  "version": "6.1.12",
  "description": "TypeScript ORM for Node.js based on Data Mapper, Unit of Work and Identity Map patterns. Supports MongoDB, MySQL, PostgreSQL and SQLite databases as well as usage with vanilla JavaScript.",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "typings": "dist/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": "./dist/index.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/mikro-orm/mikro-orm.git"
  },
  "funding": "https://github.com/sponsors/b4nan",
  "keywords": [
    "orm",
    "mongo",
    "mongodb",
    "mysql",
    "mariadb",
    "postgresql",
    "sqlite",
    "sqlite3",
    "ts",
    "typescript",
    "js",
    "javascript",
    "entity",
    "ddd",
    "mikro-orm",
    "unit-of-work",
    "data-mapper",
    "identity-map"
  ],
  "author": "Martin AdÃ¡mek",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/mikro-orm/mikro-orm/issues"
  },
  "homepage": "https://mikro-orm.io",
  "engines": {
    "node": ">= 18.12.0"
  },
  "scripts": {
    "build": "yarn clean && yarn compile && yarn copy && yarn run -T gen-esm-wrapper dist/index.js dist/index.mjs",
    "clean": "yarn run -T rimraf ./dist",
    "compile": "yarn run -T tsc -p tsconfig.build.json",
    "copy": "node ../../scripts/copy.mjs"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "dataloader": "2.2.2",
    "dotenv": "16.4.5",
    "esprima": "4.0.1",
    "fs-extra": "11.2.0",
    "globby": "11.1.0",
    "mikro-orm": "6.1.12",
    "reflect-metadata": "0.2.2"
  }
}
