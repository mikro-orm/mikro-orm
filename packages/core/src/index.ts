/**
 * @packageDocumentation
 * @module core
 */
/* istanbul ignore file */
export {
  Constructor, Dictionary, PrimaryKeyType, PrimaryKeyProp, Primary, IPrimaryKey, FilterQuery, IWrappedEntity, EntityName, EntityData, Highlighter,
  AnyEntity, EntityProperty, EntityMetadata, QBFilterQuery, PopulateOptions, Populate, Loaded, New, LoadedReference, LoadedCollection,
  GetRepository, EntityRepositoryType, MigrationObject, DeepPartial, PrimaryProperty, Cast, IsUnknown, EntityDictionary, EntityDTO, PlainObject,
} from './typings';
export * from './enums';
export * from './errors';
export * from './exceptions';
export * from './MikroORM';
export * from './entity';
export * from './events';
export * from './EntityManager';
export * from './unit-of-work';
export * from './utils';
export * from './hydration';
export * from './drivers';
export * from './connections';
export * from './platforms';
export * from './types';
export * from './naming-strategy';
export * from './metadata';
export * from './cache';
export * from './decorators';
