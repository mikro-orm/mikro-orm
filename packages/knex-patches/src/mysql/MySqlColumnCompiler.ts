// @ts-ignore
import BaseMySqlColumnCompiler from 'knex/lib/dialects/mysql/schema/mysql-columncompiler';

export class MySqlColumnCompiler extends BaseMySqlColumnCompiler {

  // we need the old behaviour to be able to add auto_increment to a column that is already PK
  increments(this: any, options = { primaryKey: true }) {
    return 'int unsigned not null auto_increment' + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

  /* istanbul ignore next */
  bigincrements(this: any, options = { primaryKey: true }) {
    return 'bigint unsigned not null auto_increment' + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

}
