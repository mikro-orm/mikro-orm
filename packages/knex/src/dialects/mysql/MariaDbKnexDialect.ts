import { MonkeyPatchable } from '../../MonkeyPatchable';

export class MariaDbKnexDialect extends MonkeyPatchable.MySqlDialect {

  get driverName() {
    return 'mariadb';
  }

  _driver() {
    return require('mariadb/callback');
  }

  validateConnection(connection: any) {
    return connection.isValid();
  }

}
