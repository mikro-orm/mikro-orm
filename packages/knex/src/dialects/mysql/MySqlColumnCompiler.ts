// @ts-ignore
import BaseMySqlColumnCompiler from 'knex/lib/dialects/mysql/schema/mysql-columncompiler';

export class MySqlColumnCompiler extends BaseMySqlColumnCompiler {

  // we need the old behaviour to be able to add auto_increment to a column that is already PK
  increments(this: any, options = { primaryKey: true, unsigned: true }) {
    const type = options.unsigned ? 'int unsigned' : 'int';
    return `${type} not null auto_increment` + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

  /* istanbul ignore next */
  bigincrements(this: any, options = { primaryKey: true, unsigned: true }) {
    const type = options.unsigned ? 'bigint unsigned' : 'bigint';
    return `${type} not null auto_increment` + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

}
