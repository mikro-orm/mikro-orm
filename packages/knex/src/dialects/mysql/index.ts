export * from './MariaDbKnexDialect';
export * from './MySqlConnection';
export * from './MySqlExceptionConverter';
export * from './MySqlSchemaHelper';
export * from './MySqlPlatform';
export * from './MySqlNativeQueryBuilder';
