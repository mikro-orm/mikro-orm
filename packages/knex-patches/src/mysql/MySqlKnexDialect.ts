import { MySqlQueryCompiler } from './MySqlQueryCompiler';
import { MySqlColumnCompiler } from './MySqlColumnCompiler';
import { MonkeyPatchable } from '../MonkeyPatchable';

export class MySqlKnexDialect extends MonkeyPatchable.MySqlDialect {

  queryCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MySqlQueryCompiler as any)(this, ...arguments);
  }

  columnCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MySqlColumnCompiler as any)(this, ...arguments);
  }

}
