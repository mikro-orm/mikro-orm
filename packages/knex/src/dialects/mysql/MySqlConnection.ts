import type { Knex } from 'knex';
import { MySqlKnexDialect } from './MySqlKnexDialect';
import { AbstractSqlConnection } from '../../AbstractSqlConnection';

export class MySqlConnection extends AbstractSqlConnection {

  override createKnex() {
    this.client = this.createKnexClient(MySqlKnexDialect as any);
    this.connected = true;
  }

  getDefaultClientUrl(): string {
    return 'mysql://root@127.0.0.1:3306';
  }

  override getConnectionOptions(): Knex.MySqlConnectionConfig {
    const ret = super.getConnectionOptions() as Knex.MySqlConnectionConfig;

    if (this.config.get('multipleStatements')) {
      ret.multipleStatements = this.config.get('multipleStatements');
    }

    if (this.config.get('forceUtcTimezone')) {
      ret.timezone = 'Z';
    }

    if (this.config.get('timezone')) {
      ret.timezone = this.config.get('timezone');
    }

    ret.supportBigNumbers = true;
    ret.dateStrings = true;

    return ret;
  }

  protected transformRawResult<T>(res: any, method: 'all' | 'get' | 'run'): T {
    if (method === 'run' && ['OkPacket', 'ResultSetHeader'].includes(res[0].constructor.name)) {
      return {
        insertId: res[0].insertId,
        affectedRows: res[0].affectedRows,
        rows: [],
      } as unknown as T;
    }

    if (method === 'get') {
      return res[0][0];
    }

    return res[0];
  }

}
