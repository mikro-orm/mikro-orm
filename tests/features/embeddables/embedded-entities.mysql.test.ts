import { Embeddable, Embedded, Entity, MikroORM, PrimaryKey, Property, ReferenceKind, wrap } from '@mikro-orm/core';
import { MySqlDriver } from '@mikro-orm/mysql';
import { mockLogger } from '../../helpers';

@Embeddable()
class Address1 {

  @Property()
  street?: string;

  @Property()
  postalCode?: string;

  @Property()
  city?: string;

  @Property()
  country?: string;

  constructor(street?: string, postalCode?: string, city?: string, country?: string) {
    this.street = street;
    this.postalCode = postalCode;
    this.city = city;
    this.country = country;
  }

}

@Embeddable()
class Address2 {

  @Property()
  street!: string;

  @Property({ nullable: true })
  postalCode?: string;

  @Property()
  city!: string;

  @Property()
  country!: string;

  constructor(street: string, city: string, country: string, postalCode?: string) {
    this.street = street;
    this.city = city;
    this.country = country;
    this.postalCode = postalCode;
  }

}

@Entity()
class User {

  @PrimaryKey()
  id!: number;

  @Embedded(() => Address1)
  address1!: Address1;

  @Embedded(() => Address2, { prefix: 'addr_', nullable: true })
  address2?: Address2;

  @Embedded({ prefix: false })
  address3: Address1 = new Address1();

  @Embedded({ object: true })
  address4: Address1 = new Address1();

  @Embedded({ object: true, nullable: true })
  address5?: Address1;

  @Property({ nullable: true })
  after?: number; // property after embeddables to verify order props in resulting schema

}

@Entity()
class UserWithCity {

  @PrimaryKey()
  id!: number;

  @Embedded(() => Address1, { object: false, prefix: false })
  address1!: Address1;

  @Property({ type: String })
  city!: string;

}

describe('embedded entities in mysql', () => {

  let orm: MikroORM<MySqlDriver>;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [User],
      dbName: `mikro_orm_test_embeddables`,
      driver: MySqlDriver,
      port: 3308,
    });
    await orm.schema.refreshDatabase();
  });

  afterAll(() => orm.close(true));

  test('metadata', async () => {
    expect(orm.getMetadata().get('Address1').embeddable).toBe(true);
    expect(orm.getMetadata().get('Address1').properties).toMatchObject({
      street: { name: 'street', type: 'string' },
      postalCode: { name: 'postalCode', type: 'string' },
      city: { name: 'city', type: 'string' },
      country: { name: 'country', type: 'string' },
    });
    expect(orm.getMetadata().get('User').properties.address1).toMatchObject({
      name: 'address1',
      kind: ReferenceKind.EMBEDDED,
      type: 'Address1',
    });
    expect(orm.getMetadata().get('User').properties.address1_street).toMatchObject({
      name: 'address1_street',
      kind: ReferenceKind.SCALAR,
      type: 'string',
    });
    expect(orm.getMetadata().get('User').properties.address2).toMatchObject({
      name: 'address2',
      kind: ReferenceKind.EMBEDDED,
      type: 'Address2',
    });
    expect(orm.getMetadata().get('User').properties.addr_street).toMatchObject({
      name: 'addr_street',
      kind: ReferenceKind.SCALAR,
      type: 'string',
      nullable: true,
    });
    expect(orm.getMetadata().get('User').properties.address3).toMatchObject({
      name: 'address3',
      kind: ReferenceKind.EMBEDDED,
      type: 'Address1',
    });
    expect(orm.getMetadata().get('User').properties.street).toMatchObject({
      name: 'street',
      kind: ReferenceKind.SCALAR,
      type: 'string',
    });
  });

  test('schema', async () => {
    await expect(orm.schema.getCreateSchemaSQL({ wrap: false })).resolves.toMatchSnapshot('embeddables 1');
    await expect(orm.schema.getUpdateSchemaSQL({ wrap: false })).resolves.toMatchSnapshot('embeddables 2');
    await expect(orm.schema.getDropSchemaSQL({ wrap: false })).resolves.toMatchSnapshot('embeddables 3');
  });

  test('persist and load', async () => {
    const user = new User();
    user.address1 = new Address1('Downing street 10', '123', 'London 1', 'UK 1');
    user.address2 = new Address2('Downing street 11', 'London 2', 'UK 2');
    user.address3 = new Address1('Downing street 12', '789', 'London 3', 'UK 3');
    user.address4 = new Address1('Downing street 13', '10', 'London 4', 'UK 4');

    const mock = mockLogger(orm, ['query']);
    await orm.em.persistAndFlush(user);
    orm.em.clear();
    expect(mock.mock.calls[0][0]).toMatch('begin');
    expect(mock.mock.calls[1][0]).toMatch('insert into `user` (`address1_street`, `address1_postal_code`, `address1_city`, `address1_country`, `addr_street`, `addr_city`, `addr_country`, `street`, `postal_code`, `city`, `country`, `address4`) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)');
    expect(mock.mock.calls[2][0]).toMatch('commit');

    const u = await orm.em.findOneOrFail(User, user.id);
    expect(mock.mock.calls[3][0]).toMatch('select `u0`.* from `user` as `u0` where `u0`.`id` = ? limit ?');
    expect(u.address1).toBeInstanceOf(Address1);
    expect(u.address1).toEqual({
      street: 'Downing street 10',
      postalCode: '123',
      city: 'London 1',
      country: 'UK 1',
    });
    expect(u.address2).toBeInstanceOf(Address2);
    expect(u.address2).toEqual({
      street: 'Downing street 11',
      postalCode: null,
      city: 'London 2',
      country: 'UK 2',
    });
    expect(u.address3).toBeInstanceOf(Address1);
    expect(u.address3).toEqual({
      street: 'Downing street 12',
      postalCode: '789',
      city: 'London 3',
      country: 'UK 3',
    });
    expect(u.address4).toBeInstanceOf(Address1);
    expect(u.address4).toEqual({
      street: 'Downing street 13',
      postalCode: '10',
      city: 'London 4',
      country: 'UK 4',
    });

    u.address2!.postalCode = '111';
    u.address4!.postalCode = '999';
    await orm.em.flush();
    expect(mock.mock.calls[4][0]).toMatch('begin');
    expect(mock.mock.calls[5][0]).toMatch('update `user` set `addr_postal_code` = ?, `address4` = ? where `id` = ?');
    expect(mock.mock.calls[6][0]).toMatch('commit');
    orm.em.clear();

    const u1 = await orm.em.findOneOrFail(User, { address1: { city: 'London 1', postalCode: '123' } });
    expect(mock.mock.calls[7][0]).toMatch('select `u0`.* from `user` as `u0` where `u0`.`address1_city` = ? and `u0`.`address1_postal_code` = ? limit ?');
    expect(u1.address1.city).toBe('London 1');
    expect(u1.address1.postalCode).toBe('123');
    const u2 = await orm.em.findOneOrFail(User, { address1: { city: /^London/ } });
    expect(mock.mock.calls[8][0]).toMatch('select `u0`.* from `user` as `u0` where `u0`.`address1_city` like ? limit ?');
    expect(u2.address1.city).toBe('London 1');
    expect(u2.address1.postalCode).toBe('123');
    expect(u2).toBe(u1);
    const u3 = await orm.em.findOneOrFail(User, { $or: [{ address1: { city: 'London 1' } }, { address1: { city: 'Berlin' } }] });
    expect(mock.mock.calls[9][0]).toMatch('select `u0`.* from `user` as `u0` where (`u0`.`address1_city` = ? or `u0`.`address1_city` = ?) limit ?');
    expect(u3.address1.city).toBe('London 1');
    expect(u3.address1.postalCode).toBe('123');
    expect(u3).toBe(u1);
    const err = `Using operators inside embeddables is not allowed, move the operator above. (property: User.address1, payload: { address1: { '$or': [ [Object], [Object] ] } })`;
    await expect(orm.em.findOneOrFail(User, { address1: { $or: [{ city: 'London 1' }, { city: 'Berlin' }] } })).rejects.toThrowError(err);
    const u4 = await orm.em.findOneOrFail(User, { address4: { postalCode: '999' } });
    expect(u4).toBe(u1);
    expect(mock.mock.calls[10][0]).toMatch('select `u0`.* from `user` as `u0` where json_extract(`u0`.`address4`, \'$.postalCode\') = ? limit ?');
  });

  test('GH issue 3063', async () => {
    const user = new User();
    orm.em.assign(user, {
      address1: { street: 'Downing street 10', postalCode: '123', city: 'London 1', country: 'UK 1' },
      address2: { street: 'Downing street 11', city: 'London 2', country: 'UK 2' },
      address3: { street: 'Downing street 12', postalCode: '789', city: 'London 3', country: 'UK 3' },
    });
    await orm.em.persistAndFlush(user);
    const r = await orm.em.fork().findOneOrFail(User, user);
    expect(r.address5).toBe(null);
  });

  test('assign', async () => {
    const user = new User();
    wrap(user).assign({
      address1: { street: 'Downing street 10', postalCode: '123', city: 'London 1', country: 'UK 1' },
      address2: { street: 'Downing street 11', city: 'London 2', country: 'UK 2' },
      address3: { street: 'Downing street 12', postalCode: '789', city: 'London 3', country: 'UK 3' },
    }, { em: orm.em });
    orm.em.assign(user, { address4: { city: '41', country: '42', postalCode: '43', street: '44' } });
    expect(user.address4).toMatchObject({ city: '41', country: '42', postalCode: '43', street: '44' });

    orm.em.assign(user, { address5: { city: '51', country: '52', postalCode: '53', street: '54' } });
    expect(user.address5).toMatchObject({ city: '51', country: '52', postalCode: '53', street: '54' });

    expect(user.address1).toBeInstanceOf(Address1);
    expect(user.address1).toEqual({
      street: 'Downing street 10',
      postalCode: '123',
      city: 'London 1',
      country: 'UK 1',
    });
    expect(user.address2).toBeInstanceOf(Address2);
    expect(user.address2).toEqual({
      street: 'Downing street 11',
      city: 'London 2',
      country: 'UK 2',
    });
    expect(user.address3).toBeInstanceOf(Address1);
    expect(user.address3).toEqual({
      street: 'Downing street 12',
      postalCode: '789',
      city: 'London 3',
      country: 'UK 3',
    });

    orm.em.assign(user, { address2: undefined });
    expect(user.address2).toBe(undefined);

    orm.em.assign(user, { address2: null });
    expect(user.address2).toBe(null);

    expect(() => {
      orm.em.assign(user, { address4: undefined });
    }).toThrow('You must pass a non-undefined value to the property address4 of entity User.');

    expect(() => {
      orm.em.assign(user, { address4: null });
    }).toThrow('You must pass a non-null value to the property address4 of entity User.');


    orm.em.assign(user, { address5: undefined });
    expect(user.address5).toBe(undefined);

    orm.em.assign(user, { address5: null });
    expect(user.address5).toBe(null);
  });

  test('should throw error with colliding definition of inlined embeddables without prefix', async () => {
    const err = `Property UserWithCity:city is being overwritten by its child property address1:city. Consider using a prefix to overcome this issue.`;
    await expect(MikroORM.init({
      entities: [Address1, UserWithCity],
      dbName: `mikro_orm_test_embeddables`,
      driver: MySqlDriver,
      port: 3308,
    })).rejects.toThrow(err);
  });

});
