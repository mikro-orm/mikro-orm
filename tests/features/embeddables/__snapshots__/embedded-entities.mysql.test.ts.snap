// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embedded entities in mysql schema: embeddables 1 1`] = `
"create table \`user\` (\`id\` int unsigned not null auto_increment primary key, \`address1_street\` varchar(255) not null, \`address1_postal_code\` varchar(255) not null, \`address1_city\` varchar(255) not null, \`address1_country\` varchar(255) not null, \`addr_street\` varchar(255) null, \`addr_postal_code\` varchar(255) null, \`addr_city\` varchar(255) null, \`addr_country\` varchar(255) null, \`street\` varchar(255) not null, \`postal_code\` varchar(255) not null, \`city\` varchar(255) not null, \`country\` varchar(255) not null, \`address4\` json not null, \`address5\` json null, \`after\` int null) default character set utf8mb4 engine = InnoDB;
"
`;

exports[`embedded entities in mysql schema: embeddables 2 1`] = `""`;

exports[`embedded entities in mysql schema: embeddables 3 1`] = `
"drop table if exists \`user\`;
"
`;
